### PR TITLE
[RHELC-1385] Satellite package backup failure fix

### PR DESCRIPTION
**Pull request by pr-watson**
This PR changes our backups to now backup the redhat.repo file when running with satellite. The reason backups were failing on satellite was due to the redhat.repo file being excluded from backups which in turn causes the .repo files not to be backed up. This change also removes the requirement for internet access to backup using hardcoded repository files and defaults to always using the the hardcoded repository files. 

A similiar failure with backing up repo files has been reported for offline conversions with satellite - [RHELC-1040](https://issues.redhat.com/browse/RHELC-1040), however this PR does not fix the issue

Jira Issues:

- [RHELC-1385](https://issues.redhat.com/browse/RHELC-1385)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] GitHub label has been added to help with Release notes 
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def run(self):
         super(DuplicatePackages, self).run()
 
         logger.task("Prepare: Check if there are any duplicate installed packages on the system")
-        output, _ = utils.run_subprocess(["/usr/bin/package-cleanup", "--dupes", "--quiet"], print_output=False)
+        output, ret = utils.run_subprocess(["/usr/bin/package-cleanup", "--dupes", "--quiet"], print_output=False)
```
Comment (SpyTec) [inline convert2rhel/actions/system_checks/duplicate_packages.py:34]: This is unrelated right? If `ret` isn't used I'd vote for `_`

File: convert2rhel/actions/pre_ponr_changes/backup_system.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
 import os
 import re
 
-from convert2rhel import actions, backup, exceptions, repo
+from convert2rhel import actions, backup, exceptions, repo, subscription
```
[the remark this answers is not in the document]
Comment (SpyTec) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:22]: Can be fixed

File: convert2rhel/backup/packages.py [hunk 3]
```diff
@@ -134,37 +134,20 @@ def enable(self):
...
-                for pkg in self.pkgs:
-                    self._backedup_pkgs_paths.append(
-                        utils.download_pkg(
-                            pkg=pkg,
-                            dest=BACKUP_DIR,
-                            set_releasever=self.set_releasever,
-                            custom_releasever=self.custom_releasever,
-                            varsdir=self.varsdir,
-                        )
-                    )
```
Comment (SpyTec) [inline convert2rhel/backup/packages.py:152]: Do we need to do this? In either case, it's being handled in #1181 so whichever gets merged first will have to fix

File: convert2rhel/actions/system_checks/duplicate_packages.py (hunk 1 above)
Comment (r0x0d) [reply convert2rhel/actions/system_checks/duplicate_packages.py:34]: So, I was discussing this with Preston last week and I asked him to check for return code in this command instead of just strings.. I think the `ret` is for him to test if there's any return code in case of failure that we can use to identify this specific problem

Comment (pr-watson) [reply convert2rhel/actions/system_checks/duplicate_packages.py:34]: Yes that was leftover from testing, i'll revert it to _

File: convert2rhel/backup/packages.py (hunk 3 above)
Comment (pr-watson) [reply convert2rhel/backup/packages.py:152]: @bocekm said that we need to use repository files regardless if there is internet or not, so I removed the internet access dependency here

Comment (r0x0d) [reply convert2rhel/backup/packages.py:152]: That's fine, either one that get merged first, we can deal with the rebase in the other PR.

File: convert2rhel/actions/pre_ponr_changes/backup_system.py [hunk 4]
```diff
@@ -87,9 +87,10 @@ def backup_yum_repos(self):
         loggerinst.info("Backing up .repo files from %s." % DEFAULT_YUM_REPOFILE_DIR)
 
         repo_files_backed_up = False
-
         for repo in os.listdir(DEFAULT_YUM_REPOFILE_DIR):
-            if repo.endswith(".repo") and repo != "redhat.repo":
+            if (repo.endswith(".repo") and repo != "redhat.repo") or (
+                subscription.should_subscribe and repo == "redhat.repo"
+            ):
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:94]: What about rewriting the loop like this? 

```python
        for repo in os.listdir(DEFAULT_YUM_REPOFILE_DIR):
            # backing up redhat.repo so repo files are properly backed up when doing satellite conversions

            if not repo.endswith(".repo"):
                # Maybe log that the file is not repo?
                return

            if not subscription.should_subscribe and repo == "redhat.repo":
                return

            repo_path = os.path.join(DEFAULT_YUM_REPOFILE_DIR, repo)
            restorable_file = RestorableFile(repo_path)
            backup.backup_control.push(restorable_file)
            repo_files_backed_up = True

```

Trying to simplify this check down to two things: 
1. If it does not end with `.repo`, we can bail out as it doesn't interest to us
2. If `should_subscribe` is false, and repo is `redhat.repo`, then we return
3. Otherwise, we back it up. By default, I believe `should_subscribe` is `False`, unless we detect the system is registered before the analysis/conversion. 


Does it make sense to you, @pr-watson and @SpyTec?

[the remark this answers is not in the document]
Comment (r0x0d) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:94]: Also, if possible, can we move the contents of this function (`backup_yum_repos`) to be inside the `run` method? I don't think this function is long enough to need an function like this, especially because `run` only calls this and exit. 

To me, it would be more clean if `run` performs the operations instead of calling the intermediate function.

[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:94]: It looks much cleaner now - I agree with this change. And I don't mind moving backup_yum_repos into the run method

[the remark this answers is not in the document]
Comment (SpyTec) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:94]: Looks a lot better, thanks @r0x0d for the suggestion!

**Comment by SpyTec** [comment]
Unit tests failing @pr-watson

**Comment by pr-watson** [comment]
> Unit tests failing @pr-watson

Working on the fixes now

File: convert2rhel/actions/pre_ponr_changes/backup_system.py [hunk 5]
```diff
@@ -75,28 +75,30 @@ class BackupRepository(actions.Action):
...
         super(BackupRepository, self).run()
 
-        self.backup_yum_repos()
-
-    def backup_yum_repos(self):
-        """Backup .repo files in /etc/yum.repos.d/ so the repositories can be restored on rollback."""
         loggerinst.info("Backing up .repo files from %s." % DEFAULT_YUM_REPOFILE_DIR)
 
-        repo_files_backed_up = False
+        if os.listdir(DEFAULT_YUM_REPOFILE_DIR) == []:
```
Comment (r0x0d) [inline convert2rhel/actions/pre_ponr_changes/backup_system.py:85]:
```suggestion
        if not os.listdir(DEFAULT_YUM_REPOFILE_DIR):
```

File: convert2rhel/actions/pre_ponr_changes/backup_system.py [hunk 6]
```diff
@@ -75,28 +75,30 @@ class BackupRepository(actions.Action):
...
+            loggerinst.info("No .repo files to back up.")
```
Comment (r0x0d) [inline convert2rhel/actions/pre_ponr_changes/backup_system.py:86]: What if we used a different message? Maybe something like: 

```suggestion
            loggerinst.info("Repository folder %s seems to be empty.", DEFAULT_YUM_REPOFILE_DIR)
```

Comment (r0x0d) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:86]: Just to differentiate from the other log message when a file does not end in .repo

File: convert2rhel/actions/pre_ponr_changes/backup_system.py [hunk 7]
```diff
@@ -75,28 +75,30 @@ class BackupRepository(actions.Action):
...
-        if not repo_files_backed_up:
-            loggerinst.info("No .repo files backed up.")
+            # backing up redhat.repo so repo files are properly backed up when doing satellite conversions
+
+            if not repo.endswith(".repo"):
+                loggerinst.info("No .repo files backed up.")
+                return
+
+            if not subscription.should_subscribe() and repo == "redhat.repo":
+                loggerinst.info("No .repo files backed up.")
```
Comment (r0x0d) [inline convert2rhel/actions/pre_ponr_changes/backup_system.py:96]: Maybe improving this message here as well

**Review by r0x0d — COMMENTED** [review]
Looks good to me. Only a couple of comments regarding the message we want to inform to the user.

**Review by r0x0d — APPROVED** [review]
Looks good!

**Comment by danmyway** [comment]
Just tested the build with a Satellite registration and the backup works fine.

File: convert2rhel/actions/pre_ponr_changes/backup_system.py [hunk 8]
```diff
@@ -75,28 +75,30 @@ class BackupRepository(actions.Action):
...
-                backup.backup_control.push(restorable_file)
-                repo_files_backed_up = True
-
-        if not repo_files_backed_up:
-            loggerinst.info("No .repo files backed up.")
+            # backing up redhat.repo so repo files are properly backed up when doing satellite conversions
+
+            if not repo.endswith(".repo"):
+                loggerinst.info("Skipping backup as file is not a repository file.")
+                return
```
[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/actions/pre_ponr_changes/backup_system.py:93]: Good catch! Let me change this

**Review by danmyway — APPROVED** [review]
Looks good to me, thanks!

**Comment by danmyway** [comment]
@pr-watson I'm merging this
@r0x0d this might resolve the issues in #1181, if you may pull the changes from @kokesak and rebase, so we can see, how that's going.
Thanks!

**Comment by pr-watson** [comment]
@danmyway thank you!
